src/util/ALE.sol

```diff
@@ -0,0 +1,770 @@
+        _borrowDola(_user, valuePlusFee, _permit, _dbrData, IMarket(_market));
+
+        if (_dbrData.dola != 0) dola.transfer(_user, _dbrData.dola);
+
+        if (_dbrData.amountIn != 0)
+            _buyDbr(_dbrData.amountIn, _dbrData.minOut, _user);
+        // Scope to avoid stack too deep error
+        {
+            uint256 balance = dola.balanceOf(address(this));
+
+            if (balance > _value + valuePlusFee)
+                dola.transfer(_user, balance - valuePlusFee);
+        }
+
+        // Refund any possible unspent fees to the sender.
+        if (address(this).balance > 0)
+            payable(_user).transfer(address(this).balance);
+
+        emit LeverageUp(
+            _market,
+            _user,
+            _value,
+            _fee,
+            collateralAmount,
+            _dbrData.dola,
+            _dbrData.amountIn
+        );
+    }
+
+    function _onFlashLoanDeleverage(
+        uint256 _value,
+        uint256 _fee,
+        bytes memory data
+    ) internal {
+        (
+            ,
+            address _user,
+            address _market,
+            uint256 _collateralAmount,
+            address _spender,
+            bytes memory _swapCallData,
+            Permit memory _permit,
+            bytes memory _helperData,
+            DBRHelper memory _dbrData
+        ) = abi.decode(
+                data,
+                (
+                    bytes32,
+                    address,
+                    address,
+                    uint256,
+                    address,
+                    bytes,
+                    Permit,
+                    bytes,
+                    DBRHelper
+                )
+            );
+
+        _repayAndWithdraw(
+            _user,
+            _value,
+            _collateralAmount,
+            _permit,
+            _dbrData,
+            IMarket(_market)
+        );
+
+        IERC20 sellToken = markets[_market].buySellToken;
+
+        // If there's a helper contract, the collateral has to be transformed
+        if (address(markets[_market].helper) != address(0)) {
+            _collateralAmount = _convertToAsset(
+                _collateralAmount,
+                _market,
+                sellToken,
+                _helperData
+            );
+        }
+
+        // Call the encoded swap function call on the contract at `swapTarget`,
+        // passing along any ETH attached to this function call to cover protocol fees.
+        // NOTE: This will swap the collateral or helperCollateral for DOLA
+        if (markets[_market].useProxy) {
+            // Approve sellToken for spender
+            sellToken.approve(_spender, 0);
+            sellToken.approve(_spender, _collateralAmount);
+            (bool success, ) = exchangeProxy.call{value: msg.value}(
+                _swapCallData
+            );
+            if (!success) revert SwapFailed();
+        }
+
+        if (address(markets[_market].helper) == address(0)) {
```

Comment:
What should happen if there is a helper contract? Helper contract pays back excess?